### PR TITLE
feat(auth): type-cacheable firestore and networkfirst adapter

### PR DESCRIPTION
**Pull request by julianpoy**
## Because

- We need to cache Contentful queries to make ourselves resilient to potential Contentful downtime.

## This pull request

- Implements a network-first cache on top of Contentful.

## Issue that this pull request solves

Closes: FXA-9099

### CODE REVIEW

File: libs/shared/contentful/src/lib/contentful.client.ts [hunk 1]
```diff
@@ -39,6 +50,26 @@ export class ContentfulClient {
...
+        );
+        const queryHash = createHash('sha256')
+          .update(JSON.stringify(args[0]))
+          .digest('hex');
+        const variableHash = createHash('sha256')
+          .update(variablesString)
+          .digest('hex');
+        return `${CONTENTFUL_QUERY_CACHE_KEY}:${queryHash}:${variableHash}`;
+      },
+      strategy: new NetworkFirstStrategy(),
```
Comment (bbangert) [inline libs/shared/contentful/src/lib/contentful.client.ts:69]: Should probably relocate this entire lambda to be a function elsewhere.

File: libs/shared/contentful/src/lib/util.ts
Comment (pdehaan) [inline libs/shared/contentful/src/lib/util.ts:1]: Does this need a copyright header?

File: libs/shared/db/firestore/jest.config.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/* eslint-disable */
```
Comment (pdehaan) [inline libs/shared/db/firestore/jest.config.ts:1]: Q1: Does this need a copyright header.
Q2: Why disable ESLint for the file?

File: libs/shared/db/firestore/src/lib/firestore-typedi-token.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+import { Firestore } from '@google-cloud/firestore';
```
Comment (pdehaan) [inline libs/shared/db/firestore/src/lib/firestore-typedi-token.ts:1]:
```suggestion
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { Firestore } from '@google-cloud/firestore';
```

File: libs/shared/db/type-cacheable/jest.config.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+/* eslint-disable */
```
Comment (pdehaan) [inline libs/shared/db/type-cacheable/jest.config.ts:1]: Same Q's as above re: copyright and disabling ESLint globally for the file.

File: libs/shared/contentful/src/lib/util.ts
Comment (julianpoy) [reply libs/shared/contentful/src/lib/util.ts:1]: Updated

File: libs/shared/db/firestore/jest.config.ts (hunk 2 above)
Comment (julianpoy) [reply libs/shared/db/firestore/jest.config.ts:1]: Q1: Updated
Q2: Nx generates it this way. I assume it doesn't conform to ESLint rules, but not sure. We're following this generator for all source, so it's this way in every new lib.

File: libs/shared/db/firestore/src/lib/firestore-typedi-token.ts (hunk 3 above)
Comment (julianpoy) [reply libs/shared/db/firestore/src/lib/firestore-typedi-token.ts:5]: Updated

File: libs/shared/db/type-cacheable/jest.config.ts (hunk 4 above)
Comment (julianpoy) [reply libs/shared/db/type-cacheable/jest.config.ts:1]: Updated and answered above :)

File: libs/shared/contentful/src/lib/contentful.client.ts [hunk 5]
```diff
@@ -58,33 +64,36 @@ export class ContentfulClient {
     return result;
   }
 
+  @FirestoreCacheable(
+    {
+      cacheKey: (args: any) => cacheKeyForQuery(args[0], args[1]),
+      strategy: new NetworkFirstStrategy(),
+      ttlSeconds: DEFAULT_FIRESTORE_CACHE_TTL,
+    },
+    CONTENTFUL_QUERY_CACHE_KEY
```
Comment (julianpoy) [inline libs/shared/contentful/src/lib/contentful.client.ts:73]: Unfortunately this can't come from DI (Contentful config) since it's provided to the decorator and won't exist in time.

File: libs/payments/capability/src/lib/capability.manager.ts [hunk 6]
```diff
@@ -40,14 +40,14 @@ export class CapabilityManager {
...
-      );
-
     const result: Record<string, string[]> = {};
 
     for (const subscribedPrice of subscribedPrices) {
+      const purchaseDetails =
+        await this.contentfulManager.getPurchaseDetailsForCapabilityServiceByPlanIds(
+          [subscribedPrice]
+        );
+
```
Comment (StaberindeZA) [inline libs/payments/capability/src/lib/capability.manager.ts:50]: Why is this change necessary? Is it what we discussed previously to limit the number of possible cacheKeys to the number of plans?

**Review by StaberindeZA — CHANGES_REQUESTED** [review]
Oh, I just noticed that `listStripePlans` is also showing up in the contentfulQueryCache. This method in `StripeHelper.allPlans` is using `@Cacheable` to cache this method to `redis`. This seems like unintended behavior?

image

image

**Comment by julianpoy** [comment]
> Oh, I just noticed that `listStripePlans` is also showing up in the contentfulQueryCache. This method in `StripeHelper.allPlans` is using `@Cacheable` to cache this method to `redis`. This seems like unintended behavior?
> 
> image
> 
> image

This should be fixed, but please give it another go. Thanks for the great catch. Unfortunately the useXYZ adapter call overrides the global CacheManager.

File: libs/shared/db/firestore/src/lib/firestore.provider.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import * as grpc from '@grpc/grpc-js';
+import { Provider } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+
+/**
+ * Creates a firestore instance from a settings object.
+ * @param config
+ * @returns A firestore instance
+ */
+export function setupFirestore(config: FirebaseFirestore.Settings) {
```
[the remark this answers is not in the document]
Comment (julianpoy) [reply libs/shared/db/firestore/src/lib/firestore.provider.ts:15]: We don't, but we should! I'll create that.

File: libs/shared/db/type-cacheable/src/lib/firestore-cacheable.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { Inject } from '@nestjs/common';
+import { Cacheable, CacheOptions } from '@type-cacheable/core';
+import { AuthFirestore, FirestoreService } from '@fxa/shared/db/firestore';
+import { useAdapter } from './type-cachable-firestore-adapter';
+import { Firestore } from '@google-cloud/firestore';
+import Container from 'typedi';
+
+export function FirestoreCacheable(
```
[the remark this answers is not in the document]
Comment (julianpoy) [reply libs/shared/db/type-cacheable/src/lib/firestore-cacheable.ts:16]: It wraps Cacheable since decorators are executed at definition, rather than at call. This injects Firestore for us on top of Cacheable.

File: libs/payments/capability/src/lib/capability.manager.ts (hunk 6 above)
Comment (julianpoy) [reply libs/payments/capability/src/lib/capability.manager.ts:50]: Correct - it's to satisfy bullet point 2 in the ticket:

> Call getPurchaseDetailsForCapabilityServiceByPlanIds for plan IDs individually, such that we greatly increase the likelihood of a cache hit

**Review by StaberindeZA — APPROVED** [review]
r+ looks good!

Lets get this into Stage!
